File: src/main/java/me/geek/tom/serverutils/commands/HomeArgumentType.java
package me.geek.tom.serverutils.commands;

import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import me.geek.tom.serverutils.Components;
import me.geek.tom.serverutils.TomsServerUtils;
import me.geek.tom.serverutils.sethome.Home;
import me.geek.tom.serverutils.sethome.HomesComponent;
import me.geek.tom.serverutils.sethome.HomesComponentImpl;
import net.minecraft.command.CommandSource;
import net.minecraft.server.command.CommandManager;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.server.network.ServerPlayerEntity;

import java.util.stream.Stream;

public class HomeArgumentType {
    public static RequiredArgumentBuilder<ServerCommandSource, String> home(String name) {
        return CommandManager.argument(name, StringArgumentType.string())
                .suggests((ctx, builder) -> {
                    ServerPlayerEntity player = ctx.getSource().getPlayer();
                    HomesComponent component = Components.HOMES.get(player);
                    boolean showAll = TomsServerUtils.homesConfig.getAllowCrossDimension();
                    Stream<String> homes = (showAll ? component.getAllHomes() : component.getAllInDimension(player.getServerWorld().getRegistryKey()))
                            .stream().map(Home::getName);

                    return CommandSource.suggestMatching(homes, builder);
                });
    }

    public static Home get(CommandContext<ServerCommandSource> ctx, String name) throws CommandSyntaxException {
        String homeName = StringArgumentType.getString(ctx, name);
        ServerPlayerEntity player = ctx.getSource().getPlayer();
        HomesComponent component = Components.HOMES.get(player);
        Home home = component.getByName(homeName);
        if (home == null)
            throw HomesComponentImpl.HOME_NOT_FOUND.create(homeName);
        return home;
    }
}

File: src/main/kotlin/commands/arguments/HomeArgumentType.kt
package me.geek.tom.serverutils.commands.arguments

import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.RequiredArgumentBuilder
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.suggestion.SuggestionsBuilder
import me.geek.tom.serverutils.Components
import me.geek.tom.serverutils.homes.Home
import me.geek.tom.serverutils.homes.HomesComponentImpl
import me.geek.tom.serverutils.homesConfig
import net.minecraft.command.CommandSource
import net.minecraft.server.command.CommandManager
import net.minecraft.server.command.ServerCommandSource

object HomeArgumentType {
    fun home(name: String): RequiredArgumentBuilder<ServerCommandSource, String> {
        return CommandManager.argument(name, StringArgumentType.string())
            .suggests { ctx: CommandContext<ServerCommandSource>, builder: SuggestionsBuilder? ->
                val player = ctx.source.player
                val component = Components.HOMES.get(player)
                val showAll = homesConfig!!.allowCrossDimension
                val homes = if (showAll)
                    component.allHomes
                else
                    component.getAllInDimension(player.serverWorld.registryKey)

                val homeNames = homes.stream()
                    .map { obj: Home -> obj.name }

                CommandSource.suggestMatching(homeNames, builder)
            }
    }

    operator fun get(ctx: CommandContext<ServerCommandSource>, name: String): Home {
        val homeName = StringArgumentType.getString(ctx, name)
        val player = ctx.source.player
        val component = Components.HOMES.get(player)
        return component.getByName(homeName) ?: throw HomesComponentImpl.HOME_NOT_FOUND.create(homeName)
    }
}
